### PR TITLE
Musical Chairs

### PR DESCRIPTION
**Pull request by bvosk**
Implemented musical chairs.

### CODE REVIEW

File: musical_chairs/lib/musical_chairs.ex [hunk 1]
```diff
@@ -1,2 +1,58 @@
...
+  defp write_results(rounds_results) do
+    File.rm(@output_file)
+    
+    rounds_results
+    |> Enum.map(&write_round/1)
+  end
+
+  defp write_round({children_eliminated, index}) do
+    round = index + 1
+    round_report = "Round ##{round}: #{children_eliminated} children eliminated\n"
```
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex]: 💀

File: musical_chairs/lib/musical_chairs.ex [hunk 2]
```diff
@@ -1,2 +1,58 @@
...
+  end
+
+  defp play_musical_chairs do
+    @input_file
+    |> parse_rounds
+    |> play_rounds
+    |> write_results
+  end
+
+  defp write_results(rounds_results) do
```
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex]: I'll typically order my functions roughly by how they are used. In this module, we are calling them roughly in order of `parse_rounds`, `play_rounds`, `write_results`. Helps the reader not have to hunt around the file so much.

File: musical_chairs/lib/musical_chairs.ex [hunk 3]
```diff
@@ -1,2 +1,47 @@
...
+  end
+
+  defp line_to_children_and_chairs(line) do
+    line
+    |> String.split
+    |> Enum.map(&String.to_integer/1)
+    |> List.to_tuple
+  end
+
+  defp play_round(round) do
```
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex]: This function is redundant. You should be able to call `eliminated_children` directly.

File: musical_chairs/lib/musical_chairs.ex [hunk 4]
```diff
@@ -1,2 +1,47 @@
...
+
+  defp play_musical_chairs do
+    File.rm(@output_file)
+
+    @input_file
+    |> File.stream!
+    |> Stream.map(&line_to_children_and_chairs/1)
+    |> Stream.map(&play_round/1)
+    |> Stream.with_index
+    |> Enum.map(&write_round/1)
```
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex:16]: Can this use `Stream.map`?

File: musical_chairs/lib/musical_chairs.ex [hunk 5]
```diff
@@ -1,2 +1,42 @@
...
+  def main(_args) do
+    play_musical_chairs
+  end
+
+  defp play_musical_chairs do
+    File.rm(@output_file)
+
+    @input_file
+    |> File.stream!
+    |> Stream.map(&line_to_children_and_chairs/1)
```
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex]: This function seems to mix levels of abstraction. Thoughts on calling this `process_round` and having it parse the data and call `eliminated_children`?

File: musical_chairs/lib/musical_chairs.ex (hunk 4 above)
Comment (bvosk) [inline musical_chairs/lib/musical_chairs.ex:16]: No I don't think so. I originally used `Stream.map`, but it doesn't work. I think you need to use `Enum` at the very end to force it to enumerate through all the lines.

File: musical_chairs/lib/musical_chairs.ex (hunk 5 above)
Comment (jsteiner) [inline musical_chairs/lib/musical_chairs.ex]: When I say mixing level of abstractions, you're reading and writing a file, but then you also have some math to calculate the `eliminated_children`, which seems lower level to me.

**Pull request merged**